src/libcore/ptr.rs

```diff
@@ -304,28 +559,61 @@ pub unsafe fn read<T>(src: *const T) -> T {
 /// Reads the value from `src` without moving it. This leaves the
 /// memory in `src` unchanged.
 ///
-/// Unlike `read`, the pointer may be unaligned.
+/// Unlike [`read`], `read_unaligned` works with unaligned pointers.
 ///
 /// # Safety
 ///
-/// Beyond accepting a raw pointer, this is unsafe because it semantically
-/// moves the value out of `src` without preventing further usage of `src`.
-/// If `T` is not `Copy`, then care must be taken to ensure that the value at
-/// `src` is not used before the data is overwritten again (e.g. with `write`,
-/// `write_bytes`, or `copy`). Note that `*src = foo` counts as a use
-/// because it will attempt to drop the value previously at `*src`.
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `src` must be [valid] for reads.
+///
+/// Like [`read`], `read_unaligned` creates a bitwise copy of `T`, regardless of
+/// whether `T` is [`Copy`].  If `T` is not [`Copy`], using both the returned
+/// value and the value at `*src` can [violate memory safety][read-ownership].
+///
+/// Note that even if `T` has size `0`, the pointer must be non-NULL and properly aligned.
+///
+/// [`Copy`]: ../marker/trait.Copy.html
+/// [`read`]: ./fn.read.html
+/// [`write_unaligned`]: ./fn.write_unaligned.html
+/// [read-ownership]: ./fn.read.html#ownership-of-the-returned-value
+/// [valid]: ../ptr/index.html#safety
 ///
 /// # Examples
 ///
-/// Basic usage:
+/// Access members of a packed struct by reference:
 ///
 /// ```
-/// let x = 12;
-/// let y = &x as *const i32;
+/// use std::ptr;
 ///
-/// unsafe {
-///     assert_eq!(std::ptr::read_unaligned(y), 12);
+/// #[repr(packed, C)]
+/// struct Packed {
+///     _padding: u8,
+///     unaligned: u32,
 /// }
+///
+/// let x = Packed {
+///     _padding: 0x00,
+///     unaligned: 0x01020304,
+/// };
+///
+/// let v = unsafe {
+///     // Take the address of a 32-bit integer which is not aligned.
+///     // This must be done as a raw pointer; unaligned references are invalid.
+///     let unaligned = &x.unaligned as *const u32;
```

Comment:
In my mental model, this creates a reference which is then cast to a raw pointer; is that incorrect?

Reply:
We don't know what this code does yet.

If your mental model is correct, then this code is UB, and there is no way to access an unaligned field of a packed struct in Rust without invoking UB. 

For this code to be correct it would need to create a raw point to the unaligned field without creating a reference. It might be possible to say that `&foo as *const T` is special and does exactly that, or we might have to add some new syntax to the language to do that.

Reply:
I think we have no choice but make this pattern of creating a raw pointer to an unaligned field legal. If this is illegal, then there just is no legal way to do this currently, and it is a fairly constrained pattern: We essentially say you must cast to a raw pointer *immediately*.

Reply:
I like that solution. Once adopted, we will have to explain in the raw pointer sections of the book that:

```rust
let x = &mut foo.x as *mut i32;
```

is not the same as 

```rust
let x = &mut foo.x;
let x = x as *mut i32;
```

and maybe the cases in which the second case is undefined behavior (if `foo.x` isn't properly aligned, maybe also if `foo.x` is uninitialized, etc.).

I wonder whether the second patter is something we should slowly start to actively warn about.

Reply:
Well, first we need a new MIR primitive for creating a raw pointer, change MIR generation to use it when immediately casting, and somewhere along the line also have an RFC for this stuff.^^

Reply:
Yes, because there's no way otherwise to do this. It'd be unreasonable to do anything else, given the design decisions up to this point around unaligned types (cc @rust-lang/lang)